src/nbl/video/CVulkanConnection.cpp

```diff
@@ -109,60 +107,65 @@ namespace nbl::video
-
-            return true;
+        };
+        auto patchDependencies = [](FeatureSetType& selectedFeatureSet, Features& actualFeaturesToEnable) -> void
+        {
+            // Vulkan Spec:
+            // If an extension is supported (as queried by vkEnumerateInstanceExtensionProperties or
+            //    vkEnumerateDeviceExtensionProperties), then required extensions of that extension must also be
+            //    supported for the same instance or physical device.
+            // -> So No need to use `insertToFeatureSetIfAvailable` because when vulkan reports an extension as supported it also has their dependancies supported
+            // TODO: No current extension needs another, except when we add DISPLAY Swapchain mode because:
+            // VK_KHR_display Requires VK_KHR_surface to be enabled
+            return;
         };
 
-        FeatureSetType selectedFeatureSet;
-        for (uint32_t i = 0u; i < requiredFeatureCount; ++i)
+
+        if(featuresToEnable.debugUtils)
         {
-            if (!insertFeatureIfAvailable(requiredFeatures[i], selectedFeatureSet))
-                return nullptr;
+            insertToFeatureSetIfAvailable(VK_EXT_DEBUG_UTILS_EXTENSION_NAME, "debugUtils");
         }
-
-        for (uint32_t i = 0u; i < optionalFeatureCount; ++i)
+        if(featuresToEnable.swapchainMode.hasFlags(E_SWAPCHAIN_MODE::ESM_SURFACE))
         {
-            if (!insertFeatureIfAvailable(optionalFeatures[i], selectedFeatureSet))
-                continue;
+            insertToFeatureSetIfAvailable(VK_KHR_SURFACE_EXTENSION_NAME, "E_SWAPCHAIN_MODE::ESM_SURFACE flag for featureName");
+#if defined(_NBL_PLATFORM_WINDOWS_)
+            insertToFeatureSetIfAvailable(VK_KHR_WIN32_SURFACE_EXTENSION_NAME, "E_SWAPCHAIN_MODE::ESM_SURFACE flag for featureName");
+#endif
         }
+        Features enabledFeatures = featuresToEnable;
+        patchDependencies(selectedFeatureSet, enabledFeatures);
 
-        const size_t totalFeatureCount = selectedFeatureSet.size() + 1ull;
-        core::vector<const char*> selectedFeatures(totalFeatureCount);
+        const size_t totalFeatureCount = selectedFeatureSet.size();
+        core::vector<const char*> extensionStringsToEnable(totalFeatureCount);
         uint32_t k = 0u;
         for (const auto& feature : selectedFeatureSet)
-            selectedFeatures[k++] = feature.c_str();
-        selectedFeatures[k++] = VK_EXT_DEBUG_UTILS_EXTENSION_NAME;
+            extensionStringsToEnable[k++] = feature.c_str();
+
+        // TODO: Figure out whether to have hard requirements for featuresToEnable and fail of not available or someway to report the available instance features/layer to user in the Features structure?!
+        // Or maybe have soft requirements and just log some warning that requested feature instance is not available? figure out the pros and cons of these choices
+        // if(!allRequestedFeaturesSupported)
+        //     return nullptr;
```

Comment:
always hard requirement, if explicitly asked for feature not available then fail to create the Instance